### PR TITLE
PR reviewer: Create inline comments on pull requests

### PR DESCRIPTION
**Pull request by bhavana-nair**
## Problem
The PR review persona needed the ability to create inline comments on pull requests to provide specific feedback on code changes.

## Solution
Added new PR comment functionality with pr_comment_tool.py that creates inline comments on GitHub pull requests
Added comprehensive unit tests to ensure reliability and maintainability

Fixed unreachable code issue in fetch_ci_failures.py.

## Changes
Code:

NEW: pr_comment_tool.py - Tool for creating inline PR comments with GitHub API integration

FIXED: fetch_ci_failures.py - Removed unreachable code

Tests:

NEW: tests/test_pr_comment_tool.py with 9 comprehensive test cases covering success scenarios, error handling, and edge cases

## Screenshots/Demo

Example: 
Screenshot 2025-06-29 at 8 32 08 PM

## Testing Instructions

Run the new test suite: python -m pytest tests/test_pr_comment_tool.py -v
Verify all existing tests still pass: python -m pytest tests/ -v

### CODE REVIEW

**Review by srdas — COMMENTED** [review]
I am able to run this persona from a fresh install and it all seems to work well till I run into a token limit, so will retry. 

Meanwhile, some comments below on each file with line numbers:

`README.md`

Line 6: Is this still the case? The PR has been updated to handle this shortcoming. Also maybe good to use `Jupyter Chat` instead of lower case without the markdown limiters. 

Line 20: What is "code smell detection"? Maybe elaborate with a sentence? 

Lines 35-47: Should not the "Feedback Loop" (node G) loop back to one  of the earlier nodes as well? Or is it as intended? Otherwise, each node can go directly to the "Final Report" node. 

Line 67, 97: Do you want to add a note on how to configure AWS credentials in the same way you have for the GH access token?

Line 73: In the implementation section, add a short description of each of the `.py` files -- what they contain and what they do. 

Lines 101-108: Elaborate a bit on each of these items.

Lines 110-113: Elaborate a bit on each of these items.


`fetch_ci_failures.py`

Line 5: redundant import?

Lines 24-25: Combine into one line of code? 


`persona.py`

For Agents, would it help to have a "description" attribute and a more detailed "role" attribute? See: https://docs.agno.com/agents/introduction


`pr_comment_tool.py`

Lines 40-41: Combine into one line of code?

**Review by srdas — COMMENTED** [review]
## 🔍 PR Review Summary

Found 5 issues that need attention. Please check the inline comments below for specific details.

**Key Areas:**
- Code quality and best practices
- Security considerations
- Documentation completeness

_Review completed by AI Assistant_

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+@tool
+def create_inline_pr_comments(
+    repo_name: str, pr_number: int, comments: List[Dict[str, Any]]
+) -> str:
+    """Create multiple inline comments on a pull request.
+
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:15]: **Type Hint Suggestion**: Consider adding return type hint in the docstring:
```python
Args:
    repo_name (str): The full name of the repository (e.g., 'owner/repo').
    pr_number (int): The number of the pull request.
    comments (List[Dict]): List of comment objects with structure...

Returns:
    str: Success message with URLs of created comments or error message.
```

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        f"create_inline_pr_comments called with repo={repo_name}, pr={pr_number}, comments={len(comments) if comments else 0}"
+    )
+    try:
+        access_token = getenv("GITHUB_ACCESS_TOKEN")
+        if not access_token:
+            return "Error: GITHUB_ACCESS_TOKEN not found"
+
+        g = Github(access_token)
+        repo = g.get_repo(repo_name)
+        pr = repo.get_pull(pr_number)
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:42]: **Rate Limiting**: Consider adding GitHub API rate limit handling here. You could:
1. Check rate limits before making API calls
2. Add exponential backoff for retries
3. Log rate limit status

Example:
```python
remaining = g.get_rate_limit().core.remaining
if remaining < len(comments) + 1:  # +1 for summary comment
    return f"Error: GitHub API rate limit too low ({remaining} remaining)"
```

File: jupyter_ai_personas/pr_review_persona/persona.py [hunk 3]
```diff
@@ -200,14 +210,49 @@ async def process_message(self, message: Message):
 
         try:
             team = self.initialize_team(system_prompt)
-            response = team.run(
-                message.body,
-                stream=False,
-                stream_intermediate_steps=False,
-                show_full_reasoning=False,
-            )
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/persona.py:213]: **Type Hints**: Consider adding type hints to improve code maintainability:
```python
async def heartbeat() -> None:
```

**Review by srdas — COMMENTED** [review]
## 🔍 PR Review Summary

Found 5 issues that need attention. Please check the inline comments below for specific details.

**Key Areas:**
- Code quality and best practices
- Security considerations
- Documentation completeness

_Review completed by AI Assistant_

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        success_count = len(comment_urls)
+        error_count = len(errors)
+
+        result = f"Posted {success_count} comments"
+        if errors:
+            result += f", {error_count} failed: {'; '.join(errors[:3])}"
+
+        return result
+    except Exception as e:
+        return f"Error: {str(e)}"
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:127]: 🔒 **Security**: The current token handling implementation may expose tokens in error logs. Please:
1. Implement token masking in error messages
2. Add secure token encryption at rest
3. Consider implementing token rotation mechanism

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    )
+    try:
+        access_token = getenv("GITHUB_ACCESS_TOKEN")
+        if not access_token:
+            return "Error: GITHUB_ACCESS_TOKEN not found"
+
+        g = Github(access_token)
+        repo = g.get_repo(repo_name)
+        pr = repo.get_pull(pr_number)
+        head_repo = pr.head.repo
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:43]: ⚠️ **Security**: Error handling could potentially leak sensitive information. Consider:
1. Adding custom error handlers to sanitize output
2. Creating standardized error response format
3. Implementing logging sanitization

**Review by srdas — COMMENTED** [review]
Change the location of the review comment to appear just below the last line of the diff. (As of now, it seems to appear in the middle of the relevant diff.)

**Review by srdas — COMMENTED** [review]
## 🔍 PR Review Summary

Found 5 issues that need attention. Please check the inline comments below for specific details.

**Key Areas:**
- Code quality and best practices
- Security considerations
- Documentation completeness

_Review completed by AI Assistant_

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    repo_name: str, pr_number: int, comments: List[Dict[str, Any]]
+) -> str:
+    """Create multiple inline comments on a pull request.
+
+    Args:
+        repo_name (str): The full name of the repository (e.g., 'owner/repo').
+        pr_number (int): The number of the pull request.
+        comments (List[Dict]): List of comment objects with the following structure:
+            [
+                {
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:21]: Consider adding type hints for the return value in the docstring for better documentation. E.g.:
```python
Returns:
    str: Success message with URLs of created comments or error message.
```

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py (hunk 2 above)
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:42]: Error messages should be more descriptive and include actionable information. Instead of just 'GITHUB_ACCESS_TOKEN not found', consider:
```python
return "Error: GITHUB_ACCESS_TOKEN environment variable not found. Please set it with a valid GitHub token."
```

File: jupyter_ai_personas/pr_review_persona/persona.py [hunk 7]
```diff
@@ -200,14 +210,49 @@ async def process_message(self, message: Message):
...
+                    await asyncio.sleep(180)
+                    if processing.is_set():
+                        self.send_message("⏳ Almost done...")
+                        await asyncio.sleep(300)
+                        if processing.is_set():
+                            self.send_message(
+                                "⏳ Taking longer than expected, please wait..."
+                            )
+
+            heartbeat_task = asyncio.create_task(heartbeat())
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/persona.py:235]: Error handling could be more specific. Instead of catching all exceptions, consider handling specific GitHub API errors, configuration errors, and network errors separately for better error reporting.

**Review by srdas — COMMENTED** [review]
## 🔍 PR Review Summary

Found 5 issues that need attention. Please check the inline comments below for specific details.

**Key Areas:**
- Code quality and best practices
- Security considerations
- Documentation completeness

_Review completed by AI Assistant_

File: jupyter_ai_personas/pr_review_persona/README.md [hunk 8]
```diff
@@ -3,88 +3,105 @@
 A specialized pull request review system designed to provide comprehensive code review and feedback in GitHub Pull Requests. This system enhances the Jupyter AI extension and leverages AWS Bedrock models for intelligent PR analysis. This persona leverages Agno Teams comprising multiple Agno Agents. [Agno](https://docs.agno.com/introduction) is a full-stack framework for building Multi-Agent Systems with memory, knowledge and reasoning.
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/README.md:3]: The README would benefit from:
1. Installation instructions
2. Basic usage examples
3. API reference section
4. Troubleshooting guide

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+@tool
+def create_inline_pr_comments(
+    repo_name: str, pr_number: int, comments: List[Dict[str, Any]]
+) -> str:
+    """Create multiple inline comments on a pull request.
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:14]: Security recommendation: Consider implementing rate limiting for GitHub API calls to prevent potential abuse and API quota exhaustion.

**Review by srdas — COMMENTED** [review]
## 🔍 PR Review Summary

Found 4 issues that need attention. Please check the inline comments below for specific details.

**Key Areas:**
- Code quality and best practices
- Security considerations
- Documentation completeness

_Review completed by AI Assistant_

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+from agno.tools import tool
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+@tool
+def create_inline_pr_comments(
+    repo_name: str, pr_number: int, comments: List[Dict[str, Any]]
+) -> str:
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:21]: Consider adding type validation for the `comments` parameter to ensure each dict contains required fields before processing. This would prevent potential runtime errors and provide better error messages.

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        pr_number (int): The number of the pull request.
+        comments (List[Dict]): List of comment objects with the following structure:
+            [
+                {
+                    "path": "path/to/file.py",  # Relative file path
+                    "position": 10,            # Line number in the file
+                    "body": "Comment text"      # The comment text
+                },
+                ...
+            ]
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:35]: Consider using a context manager (`with Github(access_token) as g:`) for better resource management and cleanup. This ensures the connection is properly closed after use.

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        access_token = getenv("GITHUB_ACCESS_TOKEN")
+        if not access_token:
+            return "Error: GITHUB_ACCESS_TOKEN not found"
+
+        g = Github(access_token)
+        repo = g.get_repo(repo_name)
+        pr = repo.get_pull(pr_number)
+        head_repo = pr.head.repo
+        commit = head_repo.get_commit(pr.head.sha)
+
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:52]: Consider adding rate limiting handling for GitHub API calls. With multiple comments, you might hit rate limits. Add retry logic or rate limit checks to handle this gracefully.

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+                if not all(key in comment_data for key in ["path", "position", "body"]):
+                    logger.warning(f"Comment {i + 1}: Missing required fields")
+                    errors.append(f"Comment {i + 1}: Missing required fields")
+                    continue
+
+                file_path = comment_data["path"]
+                line_number = comment_data["position"]
+                logger.debug(f"Comment {i + 1}: Targeting {file_path}:{line_number}")
+
+                if file_path not in pr_files:
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:91]: The error handling could be more specific. Instead of catching all exceptions, consider catching specific GitHub exceptions (GithubException) separately from general exceptions to provide more detailed error messages.

**Review by srdas — COMMENTED** [review]
## 🔍 PR Review Summary

Found 4 issues that need attention. Please check the inline comments below for specific details.

**Key Areas:**
- Code quality and best practices
- Security considerations
- Documentation completeness

_Review completed by AI Assistant_

File: jupyter_ai_personas/pr_review_persona/persona.py [hunk 14]
```diff
@@ -200,14 +210,49 @@ async def process_message(self, message: Message):
...
-                show_full_reasoning=False,
-            )
 
-            self.send_message(response.content)
+            # Add periodic heartbeat messages during processing
+            import asyncio
+            import threading
+
+            # Flag to stop heartbeat when done
+            processing = threading.Event()
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/persona.py]: Consider parameterizing the heartbeat intervals (120s, 180s, 300s) as class constants or configuration values. This would make them easier to adjust without changing code.

File: jupyter_ai_personas/pr_review_persona/persona.py [hunk 15]
```diff
@@ -200,14 +210,49 @@ async def process_message(self, message: Message):
...
+            async def heartbeat():
+                await asyncio.sleep(120)
+                if processing.is_set():
+                    self.send_message("⏳ Still processing large PR...")
+                    await asyncio.sleep(180)
+                    if processing.is_set():
+                        self.send_message("⏳ Almost done...")
+                        await asyncio.sleep(300)
+                        if processing.is_set():
+                            self.send_message(
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/persona.py]: The heartbeat task cancellation logic could race with heartbeat execution. Consider adding a proper async cancellation handler using `try/except asyncio.CancelledError` to ensure clean shutdown.

File: jupyter_ai_personas/pr_review_persona/persona.py [hunk 16]
```diff
@@ -200,14 +210,49 @@ async def process_message(self, message: Message):
...
-            response = team.run(
-                message.body,
-                stream=False,
-                stream_intermediate_steps=False,
-                show_full_reasoning=False,
-            )
 
-            self.send_message(response.content)
+            # Add periodic heartbeat messages during processing
+            import asyncio
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/persona.py:235]: The processing flag is shared between async tasks. Consider using asyncio.Event() instead of threading.Event() for better async safety and consistency.

File: jupyter_ai_personas/pr_review_persona/persona.py [hunk 17]
```diff
@@ -200,14 +210,49 @@ async def process_message(self, message: Message):
...
+                                "⏳ Taking longer than expected, please wait..."
+                            )
+
+            heartbeat_task = asyncio.create_task(heartbeat())
+
+            try:
+                response = await asyncio.to_thread(
+                    team.run,
+                    message.body,
+                    stream=False,
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/persona.py:266]: Error handling in process_message should catch and handle specific exceptions like GithubException separately from general exceptions. This would provide better error reporting for different failure modes.

**Review by srdas — COMMENTED** [review]
## 🔍 PR Review Summary

Found 4 issues that need attention. Please check the inline comments below for specific details.

**Key Areas:**
- Code quality and best practices
- Security considerations
- Documentation completeness

_Review completed by AI Assistant_

File: tests/test_pr_comment_tool.py [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+    ]
+
+
+def test_create_inline_pr_comments_no_token():
+    """Test error when GitHub token is not available."""
+    with patch("os.getenv", return_value=None):
+        result = create_inline_pr_comments_logic("owner/repo", 123, [])
+        assert result == "Error: GITHUB_ACCESS_TOKEN not found"
+
+
```
Comment (srdas) [inline tests/test_pr_comment_tool.py:68]: Consider adding test cases for concurrent PR comment creation to validate the async safety of the implementation.

File: tests/test_pr_comment_tool.py [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+        mock_pr.create_issue_comment.assert_called_once()
+        assert mock_pr.create_comment.call_count == 2
+
+
+def test_create_inline_pr_comments_empty_list():
+    """Test with empty comments list."""
+    mock_commit = Mock()
+
+    mock_head_repo = Mock()
+    mock_head_repo.get_commit.return_value = mock_commit
```
Comment (srdas) [inline tests/test_pr_comment_tool.py:108]: Add test cases for handling GitHub API rate limiting scenarios to ensure proper error handling and retry logic.

File: tests/test_pr_comment_tool.py [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+    mock_comment.html_url = "https://github.com/owner/repo/pull/123#discussion_r456"
+
+    mock_commit = Mock()
+
+    mock_head_repo = Mock()
+    mock_head_repo.get_commit.return_value = mock_commit
+
+    mock_pr = Mock()
+    mock_pr.head.sha = "abc123"
+    mock_pr.head.repo = mock_head_repo
```
Comment (srdas) [inline tests/test_pr_comment_tool.py:154]: Consider adding a test case for comment creation with Unicode/special characters to ensure proper text encoding handling.

File: tests/conftest.py [hunk 21]
```diff
@@ -58,6 +77,25 @@ async def set_local_state_field(self, *args, **kwargs):
         return
 
 
+# Mock asyncio operations
```
Comment (srdas) [inline tests/conftest.py:80]: The mock_asyncio_operations fixture could be more comprehensive. Consider adding test cases for timeouts and cancellation scenarios.

**Review by srdas — CHANGES_REQUESTED** [review]
Thanks for this very useful PR. I tested in a new environment and it works well. I tested it on this very PR, and have left the comments in as part of the test. The test summary is shown here: 
image

The throttling issue that appeared earlier has been resolved with changing the LLM to `Claude Sonnet 3.5 v2`.  
 
All suggested changes have been made. A few remain for consideration, if possible, see suggestions made by the Persona: 
1. In `persona.py` -- Consider using asyncio.Event() instead of threading.Event()?
2. Add in the heartbeat intervals as constants. 
3.  More detailed error handling, only if possible.

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py (hunk 12 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:52]: Using PyGithub insted of GitHub API to avoid hitting rate limits.

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py (hunk 13 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/pr_comment_tool.py:91]: Handled this.

File: jupyter_ai_personas/pr_review_persona/persona.py (hunk 16 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/persona.py:235]: Replaced threading with asyncio.

File: jupyter_ai_personas/pr_review_persona/persona.py (hunk 17 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/persona.py:266]: Implemented GithubException handling separately.

File: jupyter_ai_personas/pr_review_persona/persona.py (hunk 14 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/persona.py]: Stored heartbeat intervals as constants. It improves code readability.

File: jupyter_ai_personas/pr_review_persona/persona.py (hunk 15 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/persona.py]: Done

File: jupyter_ai_personas/pr_review_persona/README.md [hunk 22]
```diff
@@ -3,88 +3,105 @@
...
 
 ## Work In Progress:
-Current capabilities are limited and can provide reviews as response in jupyter chat and not directly on Github PR. 
-Response is delayed as the persona runs multiple verifications and test, wait time is between to 30 sec to 1 min.
+Response is delayed as the persona runs multiple verifications and tests, wait time is between 2 min to 5 min.
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/README.md]: Do we need to have a WIP -- this PR is complete, right? Maybe call it a "Run time note" (or something like that)?

File: jupyter_ai_personas/pr_review_persona/README.md (hunk 8 above)
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/README.md]: Add a comma after knowledge (conform to the Oxford comma rule).

File: jupyter_ai_personas/pr_review_persona/persona.py [hunk 23]
```diff
@@ -58,16 +67,22 @@ def initialize_team(self, system_prompt):
...
                 "   - Error handling and edge cases",
+                "4. MUST create inline comments for issues found:",
+                "   - For each code issue, IMMEDIATELY call create_inline_pr_comments",
+                "   - Use exact file paths from PR changes",
+                "   - Use line numbers from the diff",
+                "   - Do not just mention issues - CREATE the comments",
             ],
             tools=[
-                PythonTools(),
+                # PythonTools(),
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/persona.py]: Do you still need this line? Else remove `PythonTools()`

File: jupyter_ai_personas/pr_review_persona/persona.py [hunk 24]
```diff
@@ -99,7 +114,7 @@ def initialize_team(self, system_prompt):
                 "4. Check for insecure direct object references",
             ],
             tools=[
-                PythonTools(),
+                # PythonTools(),
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/persona.py]: Remove this commented code?

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+    Returns:
+        str: Success message with URLs of created comments or error.
+    """
+    logger.debug(
+        f"create_inline_pr_comments called with repo={repo_name}, pr={pr_number}, comments={len(comments) if comments else 0}"
+    )
+    try:
+        access_token = getenv("GITHUB_ACCESS_TOKEN")
+        if not access_token:
+            return "Error: GITHUB_ACCESS_TOKEN not found"
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py]: 1. Does this throw an error or gracefully inform the user in the chat?
2. Maybe give a more detailed error message, asking the user to set up the classic token and detailing it a bit (a couple of sentences).

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+
+        g = Github(access_token)
+        repo = g.get_repo(repo_name)
```
Comment (srdas) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py]: Can this be collapsed into a single line of code by eliminating the variable `g`? (same in other places in the code)

**Review by srdas — CHANGES_REQUESTED** [review]
I have a few more small comments, but it looks good otherwise.

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py (hunk 25 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/pr_comment_tool.py]: Since its the agent calling the tool, this error would go to the agent, which would send out a proper message to the user. But I will add the classic token detail to this error.

File: jupyter_ai_personas/pr_review_persona/README.md (hunk 22 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/README.md]: Sure

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+from os import getenv
+from agno.tools import tool
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+@tool
+def create_inline_pr_comments(
+    repo_name: str, pr_number: int, comments: List[Dict[str, Any]]
```
Comment (andrii-i) [inline jupyter_ai_personas/pr_review_persona/pr_comment_tool.py]: To improve type safety and code clarity instead of using `Any` in `List[Dict[str, Any]]` for the comments parameter and relying on people reading and following comments, consider defining a `TypedDict`. This will help other developers understand the code easier and catch type errors early https://peps.python.org/pep-0589/

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py (hunk 26 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/pr_comment_tool.py]: This is changes.

File: jupyter_ai_personas/pr_review_persona/pr_comment_tool.py (hunk 27 above)
Comment (bhavana-nair) [reply jupyter_ai_personas/pr_review_persona/pr_comment_tool.py]: This is implemented and pushed along with latest commit.

**Review by srdas — APPROVED** [review]
Thanks for making the suggested changes! LGTM.

**Comment by srdas** [comment]
Nice work!! TY for the contribution!
